File: backend/src/controllers/categories.controller.ts
import Categorie from '../classes/Categories';
import { error, success } from '../utils/functions';
import type { Request, Response } from 'express';
import type { App } from '@app';

export default class CategoriesController {
  app: App;
  Categories: Categorie;
  constructor(app: App) {
    this.app = app;
    this.Categories = new Categorie(app);
  }
  getAll(_: Request, res: Response): void {
    this.Categories.getAll()
      .then((result: any) => res.status(200).json(success(result)))
      .catch((err: Error) => res.status(500).json(error(err.message)));
  }
  addMainCategory(req: Request, res: Response): void {
    this.Categories.addMainCategory(req.body.name, req.body.description, req.body.order, req.body.path, req.body.icon)
      .then(c => res.status(201).json(success(c)))
      .catch(err => res.status(500).json(error(err.message)));
  }
  addSubCategory(req: Request, res: Response): void {
    this.Categories.addSubCategory(
      req.body.name,
      req.body.description,
      req.body.order,
      req.body.path,
      req.body.icon,
      req.body.parent_category,
    )
      .then(c => res.status(201).json(success(c)))
      .catch(err => res.status(500).json(error(err.message)));
  }
  updateMainCategory(req: Request, res: Response) {
    this.Categories.updateMainCategory(
      req.params.id as string,
      req.body.name,
      req.body.description,
      req.body.order,
      req.body.path,
      req.body.icon,
    )
      .then(() => res.status(200).json(success('success')))
      .catch(err => res.status(500).json(error(err.message)));
  }
  updateSubCategory(req: Request, res: Response) {
    this.Categories.updateSubCategory(
      req.params.id as string,
      req.body.name,
      req.body.description,
      req.body.order,
      req.body.path,
      req.body.icon,
      req.body.parent_category,
    )
      .then(() => res.status(200).json(success('success')))
      .catch(err => res.status(500).json(error(err.message)));
  }
  deleteMainCategory(req: Request, res: Response) {
    this.Categories.deleteMainCategory(req.params.id as string)
      .then(() => res.status(200).json(success('success')))
      .catch(err => res.status(500).json(error(err.message)));
  }
  deleteSubCategory(req: Request, res: Response) {
    this.Categories.deleteSubCategory(req.params.id as string)
      .then(() => res.status(200).json(success('success')))
      .catch(err => res.status(500).json(error(err.message)));
  }
}

File: backend/controllers/categories.controller.go
package controllers

import (
	"alexandrie/app"
	"alexandrie/models"
	"alexandrie/utils"
	"errors"
	"net/http"

	"github.com/gin-gonic/gin"
)

type CategoriesController interface {
	GetCategories(c *gin.Context) (int, any)
	CreateCategory(c *gin.Context) (int, any)
	UpdateCategory(c *gin.Context) (int, any)
	DeleteCategory(c *gin.Context) (int, any)
}

func NewCategoriesController(app *app.App) CategoriesController {
	return &Controller{
		app: app,
	}
}

// Get categories
// @Summary Get all categories
// @Method GET
// @Router /categories [get]
// @Security Authenfification: Auth
// @Success 200 {object} Success([]models.Category)
// @Failure 400 {object} Error
// @Failure 401 {object} Error
func (ctr *Controller) GetCategories(c *gin.Context) (int, any) {
	id, err := utils.SelfOrPermission(c, utils.ADMINISTRATOR, "userId")
	if err != nil {
		return http.StatusUnauthorized, err
	}
	categories, err := ctr.app.Services.Category.GetAllCategories(id)
	if err != nil {
		return http.StatusInternalServerError, err
	}
	return http.StatusOK, categories
}

// Create category
// @Summary Create a new category
// @Method POST
// @Router /categories [post]
// @Security Authenfification: Auth
// @Param category body models.Category true "Category"
// @Success 201 {object} Success(models.Category)
// @Failure 400 {object} Error
// @Failure 401 {object} Error
func (ctr *Controller) CreateCategory(c *gin.Context) (int, any) {
	var category models.Category
	if err := c.ShouldBind(&category); err != nil {
		return http.StatusBadRequest, errors.New("invalid request payload")
	}

	userId, err := utils.GetUserIdCtx(c)
	if err != nil {
		return http.StatusUnauthorized, err
	}
	category = models.Category{
		Id:          ctr.app.Snowflake.Generate(),
		Name:        category.Name,
		Icon:        category.Icon,
		Order:       category.Order,
		Role:        category.Role,
		WorkspaceId: category.WorkspaceId,
		ParentId:    category.ParentId,
		AuthorId:    userId,
	}

	if err := ctr.app.Services.Category.CreateCategory(&category); err != nil {
		return http.StatusInternalServerError, err
	}
	return http.StatusCreated, category
}

// Update category
// @Summary Update a category
// @Method PUT
// @Router /categories/{id} [put]
// @Security Authenfification: Auth
// @Param id path string true "Category ID"
// @Param category body models.Category true "Category"
// @Success 200 {object} Success(models.Category)
// @Failure 400 {object} Error
// @Failure 401 {object} Error
func (ctr *Controller) UpdateCategory(c *gin.Context) (int, any) {
	categoryid, err := utils.GetIdParam(c, c.Param("id"))
	if err != nil {
		return http.StatusBadRequest, err
	}
	dbCategory, err := ctr.app.Services.Category.GetCategory(categoryid)
	if err != nil {
		return http.StatusInternalServerError, err
	}
	err = utils.RessourceAccess(c, dbCategory.AuthorId)
	if err != nil {
		return http.StatusUnauthorized, err
	}
	category := &models.Category{}
	if err := c.ShouldBind(category); err != nil {
		return http.StatusBadRequest, err
	}
	category = &models.Category{
		Id:          categoryid,
		Name:        category.Name,
		Icon:        category.Icon,
		Order:       category.Order,
		Role:        category.Role,
		WorkspaceId: category.WorkspaceId,
		ParentId:    category.ParentId,
		AuthorId:    dbCategory.AuthorId,
	}

	err = ctr.app.Services.Category.UpdateCategory(category)
	if err != nil {
		return http.StatusInternalServerError, err
	}
	return http.StatusOK, category
}

// Delete category
// @Summary Delete a category
// @Method DELETE
// @Router /categories/{id} [delete]
// @Security Authenfification: Auth
// @Param id path string true "Category ID"
// @Success 204 {object} Success
// @Failure 400 {object} Error
// @Failure 401 {object} Error

// Delete document
// @Summary Delete document
// @Method DELETE
// @Router /documents/{id} [delete]
// @Security Authenfification: Auth
// @Param id path string true "Document ID"
// @Success 200 {object} Success(models.Document)
// @Failure 400 {object} Error
// @Failure 401 {object} Error
func (ctr *Controller) DeleteCategory(c *gin.Context) (int, any) {
	categoryId, err := utils.GetIdParam(c, c.Param("id"))
	if err != nil {
		return http.StatusBadRequest, err
	}
	db_category, err := ctr.app.Services.Category.GetCategory(categoryId)
	if err != nil {
		return http.StatusInternalServerError, err
	}
	err = utils.RessourceAccess(c, db_category.AuthorId)
	if err != nil {
		return http.StatusUnauthorized, err
	}
	err = ctr.app.Services.Category.DeleteCategory(categoryId)
	if err != nil {
		return http.StatusInternalServerError, err
	}
	return http.StatusOK, db_category
}
